### ISSUE TITLE
svc deploy after update to copilot 1.21.1 (from 1.15.0) failing regarding load balancer update 

### ISSUE THREAD

**Issue by rmarops** [aws/copilot-cli#4011]
Extracted error (see attachment)
`Resource handler returned message: "Invalid request provided: UpdateService error: Health check grace period is only valid for services configured to use load balancers (Service: AmazonECS; Status Code: 400; Error Code: InvalidParameterException; Request ID: xyz; Proxy: null)" (RequestToken: zyx, HandlerErrorCode: InvalidRequest)`

[udpate.txt](https://github.com/aws/copilot-cli/files/9601132/udpate.txt)

appears to be similar in nature to aws-cdk reported issue: https://github.com/aws/aws-cdk/issues/19842

**Comment by rmarops** [comment]
seems related: https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/1313

right now copilot is hung on the rollback.  hoping for a potential workaround for deploy issue.

**Comment by efekarakus** [comment]
Hi @rmarops !

I think I get what's going on 💭 On Mar 2022, ECS added support to update an existing service with a load balancer (https://aws.amazon.com/about-aws/whats-new/2022/03/amazon-ecs-service-api-updating-elastic-load-balancers-service-registries-tag-propagation-ecs-managed-tags/). When a [`healthcheck.grace_period`](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#http-healthcheck-grace-period) is specified at the same as a load balancer to an existing service, then there is a race-condition and validation fails saying that the service doesn't yet have a load balancer.

I believe a workaround for now is to:
1. First add the `http` field without a `healthcheck.grace_period` and `copilot deploy`
2. Then add the field and run `copilot deploy` again.

Hope this helps!

**Comment by rmarops** [comment]
I think you are suggesting for me to set a manifest setting for the service, but I dont set this value anywhere.   As a side note, I searched the cloudformation templates that are used by copilot and it doesn't look like the equivalent (anything with grace in the name) is being set there either.  Perhaps there is an implied default.  

The AWS-CDK guy in the link above, used `(service.node.defaultChild as CfnService).healthCheckGracePeriodSeconds = undefined;` to work around the issue from his specific context.

My thoughts are I may need to determine the cloudformation equivalent and tweak the template.

Appreciate the response.

**Comment by efekarakus** [comment]
Interesting, I assumed that this error happened after enabling the `http` field for a Backend Service. It sounds like I was completely wrong :) would you mind sharing your manifest file that resulted in a failed deployment?

**Comment by rmarops** [comment]
You are on the right track.  I commented the entire http section and it succeeds, but the moment the config is as per below if fails with the same error.  My assumption is the default value for grace_period is being applied.

```
name: api
type: Backend Service

http:
  # Requests to this path will be forwarded to your service.
  # To match all requests you can use the "/" path.
  path: '/api'
  # You can specify a custom health check path. The default is "/".
  # healthcheck: '/
# Configuration for your containers and service.
image:
  # Docker build arguments. For additional overrides: https://aws.github.io/copilot-cli/docs/manifest/backend-service/#image-build
  build: Dockerfile
  # Port exposed through your container to route traffic to it.
  port: 5001

cpu: 1024 # Number of CPU units for the task.
memory: 2048 # Amount of memory in MiB used by the task.
count: 1 # Number of tasks that should be running in your service.
exec: true # Enable running commands in your container.

# <environment section removed>
```

**Comment by efekarakus** [comment]
Hi @rmarops !

You're right a `grace_period` is applied by default which forces the error to occur.
I tried also deploying with:
```yaml
http:
  healthcheck:
     grace_period: 0s
```
However, the same error happens :(

I can't think of any mitigation today besides creating a brand new backend service with the `http` field set before deploying the service.

We'll explore a workaround in the mean time to avoid setting the grace_period if the value is 0 so that at least we can do the two-step deployment.

**Comment by efekarakus** [comment]
I just sent out a PR (https://github.com/aws/copilot-cli/pull/4015) allowing to unset the GracePeriod value when the zero-value `0s` is written in the manifest.
This matches with the default ECS behavior:
> If your service has a load balancer defined and you don't specify a health check grace period value, the default value of 0 is used.

I tested the change and once the following is written and deployed the load balancer gets added to an existing service:
```
http:
  healthcheck:
     grace_period: 0s
```

**Comment by rmarops** [comment]
I tried to build copilot (run make with go installed) and it just hangs on the go command.  any estimate on next binary release?

```
$ make
Installing custom resource dependencies
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated sane@4.1.0: some dependency vulnerabilities fixed, support for node < 10 dropped, and newer ECMAScript syntax/features added
npm WARN deprecated querystring@0.2.0: The querystring API is considered Legacy. new code should use the URLSearchParams API instead.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uuid@3.3.2: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.

added 904 packages, and audited 905 packages in 5s

71 packages are looking for funding
  run `npm fund` for details

2 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
Packaging custom resources to templates/custom-resources

> ecs_cli_custom_resources@0.0.1 package
> minify lib -d ../internal/pkg/template/templates/custom-resources --builtIns false

CGO_ENABLED=0 go build -ldflags "-X github.com/aws/copilot-cli/internal/pkg/version.Version=v1.21.1-29-g2efa0171 -X github.com/aws/copilot-cli/internal/pkg/cli.binaryS3BucketPath=https://ecs-cli-v2-release.s3.amazonaws.com" -o ./bin/local/copilot ./cmd/copilot

```

**Comment by efekarakus** [comment]
We're planning to release this week (more specifically if we get our PRs in tomorrow morning)! 
That is strange though that it would hang, you can also try `make docker-release`

**Comment by efekarakus** [comment]
@rmarops This fix is now released :) in [v1.22](https://github.com/aws/copilot-cli/releases/tag/v1.22.0) 🚀 !

For the blog post: https://aws.github.io/copilot-cli/blogs/release-v122/
Release notes: https://github.com/aws/copilot-cli/releases/tag/v1.22.0
